src/search_iterator.cpp

```diff
@@ -256,4 +256,178 @@ SearchIterator::pointer SearchIterator::operator->() const {
     return &**this;
 }
 
+SuggestionIterator::~SuggestionIterator() = default;
+SuggestionIterator::SuggestionIterator(SuggestionIterator&& it) = default;
+SuggestionIterator& SuggestionIterator::operator=(SuggestionIterator&& it) = default;
+
+SuggestionIterator::SuggestionIterator()
+  : mp_rangeIterator(nullptr),
+    mp_searchIterator(nullptr)
+{}
+
+SuggestionIterator::SuggestionIterator(RangeIterator rangeIterator)
+  : mp_rangeIterator(std::unique_ptr<RangeIterator>(new RangeIterator(rangeIterator))),
+    mp_searchIterator(nullptr)
+{}
+
+SuggestionIterator::SuggestionIterator(SearchIterator searchIterator)
+  : mp_rangeIterator(nullptr),
+    mp_searchIterator(std::unique_ptr<SearchIterator>(new SearchIterator(searchIterator)))
+{}
+
+SuggestionIterator::SuggestionIterator(const SuggestionIterator& it)
+    : mp_rangeIterator(nullptr), mp_searchIterator(nullptr)
+{
+    if (it.mp_searchIterator) mp_searchIterator = std::unique_ptr<SearchIterator>(new SearchIterator(*it.mp_searchIterator));
+    else if (it.mp_rangeIterator) mp_rangeIterator = std::unique_ptr<RangeIterator>(new RangeIterator(*it.mp_rangeIterator));
+}
+
+SuggestionIterator & SuggestionIterator::operator=(const SuggestionIterator& it) {
+    *mp_rangeIterator = *it.mp_rangeIterator;
+    *mp_searchIterator = *it.mp_searchIterator;
+    return *this;
+}
+
+bool SuggestionIterator::operator==(const SuggestionIterator& it) const {
+    if (mp_rangeIterator && it.mp_rangeIterator) {
+        return (*mp_rangeIterator == *it.mp_rangeIterator);
+    } else if (mp_searchIterator && it.mp_searchIterator) {
+        return (*mp_searchIterator == *it.mp_searchIterator);
+    }
+    return false;
+}
+
+bool SuggestionIterator::operator!=(const SuggestionIterator& it) const {
+    return ! (*this == it);
+}
+
+SuggestionIterator& SuggestionIterator::operator++() {
+    if (!mp_rangeIterator && !mp_searchIterator) {
+        return *this;
+    } else if (!mp_rangeIterator) {
+        ++(*mp_searchIterator);
+    } else {
+        ++(*mp_rangeIterator);
+    }
+    return *this;
```

Comment:
We could rewrite this :
```cpp
if (mp_searchIterator) {
	++(*mp_searchIterator);
} else if (mp_rangeIterator) {
	++(*mp_rangeIterator);
}
return *this;
```

First, we have less tests. And second, "positive" conditions are better than negative ones.

If we really want to keep a defensive programming, we should write it this way :
```cpp
if (!mp_rangeIterator && !mp_searchIterator) {
	return *this;
}

if (mp_searchIterator) {
	++(*mp_searchIterator);
} else {
    ++(*mp_rangeIterator);
}

return *this;
```

Reply:
Thanks for the tip! This looks much better.